ui-react/packages/atlasmap/src/UI/ExpressionContent.tsx

```diff
@@ -1,51 +1,27 @@
-import { FunctionComponent } from "react";
+import { FunctionComponent, useEffect, useCallback } from "react";
 import React from "react";
 import { Tooltip, Form, FormGroup } from "@patternfly/react-core";
-import { Observable } from "rxjs";
+import { Subscription } from "rxjs";
```

Comment:
Picking this line but commenting on the whole file. This file is heavily coupled with the RxJS store, which is bad because it makes testing this component from a pure UI perspective very difficult. 

Since this component is in the `UI` folder it shouldn't really know or care about how things work under the hood. It should care about how things are displayed to the user and how the user can interact with the component, while exposing the necessary callbacks for the consumer of this component to do glue it to the business logic.

An handy way to understand if a component is too coupled is writing a test and/or a story for it. If you have to bring into the story too many bits it's a signal that the component is too coupled.

Reply:
What's evident about this file is that it has little to do with the UI - it's just a DIV in a react form that calls a typeahead search select.  It is likely difficult to test - it was difficult to write - but it doesn't violate any separation of duties set out in the architecture and it is providing a very specific purpose.  I think it's isolated well and attempting to refactor it into multiple abstractions to satisfy testability is wrong.

Reply:
Then the UI component should be thin and expose the callbacks needed to wire it to the code that knows about the logic. Think about how e.g. the `select` html component works, it's generic and exposes callbacks you can use to do your own specialized work.

If this UI component should only show some text and open a drawer when a special key (`@`) is keyed down, it should do only that.

All the rest should stay away, in the `Atlasmap` folder I'd say. And IMHO there should be no need for these function to know about the input field or the html that's been used to render the widget. If the interface between the logic and the ui uses basic types like `string`, then they are decoupled from each other allowing us to change the UI in the future without worrying about breaking the logic.